agent/agent.go

```diff
@@ -528,153 +571,117 @@ L:
 	if err != nil {
-		resp.err = trace.Wrap(err)
-	} else {
-		defer client.Close()
-		var status *pb.NodeStatus
-		status, err = client.LocalStatus(ctx)
-		if err != nil {
-			resp.err = trace.Wrap(err)
-		} else {
-			resp.NodeStatus = status
-		}
-	}
-	select {
-	case respc <- resp:
-	case <-r.done:
-	}
-}
-
-// timelineUpdateLoop periodically updates the status timeline.
-func (r *agent) timelineUpdateLoop(ctx context.Context) {
-	ticker := r.Clock.NewTicker(statusUpdateTimeout)
-	defer ticker.Stop()
-
-	for range ticker.Chan() {
-		if utils.IsContextDone(ctx) {
-			log.Info("Timeline update loop is stopping.")
-			return
-		}
-		if err := r.updateTimeline(ctx); err != nil {
-			log.WithError(err).Error("Failed to update timeline.")
-		}
+		return trace.Wrap(err)
 	}
-}
 
-// updateTimeline records any status changes to the timeline.
-func (r *agent) updateTimeline(ctx context.Context) error {
-	if err := r.updateLocalChanges(ctx); err != nil {
-		return trace.Wrap(err, "error updating local timeline changes")
-	}
-	r.memberIndex++
-	member, err := r.selectMember(int(r.memberIndex))
+	events, err := r.LocalTimeline.GetEvents(ctx, nil)
 	if err != nil {
-		return nil
-	}
-	if err := r.updateRemoteChanges(ctx, member); err != nil {
-		return trace.Wrap(err, "error updating remote timeline changes")
+		return trace.Wrap(err)
 	}
-	return nil
-}
 
-// updateLocalChanges updates the timeline with any new status changes found on
-// the local node.
-func (r *agent) updateLocalChanges(ctx context.Context) error {
-	ctxUpdate, cancel := context.WithTimeout(ctx, updateTimelineTimeout)
-	defer cancel()
-
-	if err := r.Timeline.RecordStatus(ctxUpdate, r.LocalStatus()); err != nil {
-		return trace.Wrap(err)
+	// TODO: async
+	for _, member := range members {
+		if role, ok := member.Tags["role"]; ok && Role(role) == RoleMaster {
+			if err := r.notifyMaster(ctx, &member, events); err != nil {
+				log.WithError(err).Warnf("Failed to notify %s of new timeline events.", member.Name)
+			}
+		}
 	}
 
 	return nil
 }
 
-// updateRemoteChanges updates the timeline with events from another member in
-// the cluster.
-func (r *agent) updateRemoteChanges(ctx context.Context, member serf.Member) error {
-	ctxQuery, cancel := context.WithTimeout(ctx, r.statusQueryReplyTimeout)
-	defer cancel()
-	timeline, err := r.getTimelineFrom(ctxQuery, member)
+// notifyMaster push new timeline events to the specified member.
+func (r *agent) notifyMaster(ctx context.Context, member *serf.Member, events []*pb.TimelineEvent) error {
+	client, err := r.dialRPC(ctx, member)
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	defer client.Close()
 
-	ctxUpdate, cancel := context.WithTimeout(ctx, updateTimelineTimeout)
-	defer cancel()
-	if err := r.Timeline.RecordTimeline(ctxUpdate, timeline); err != nil {
+	// Filter events by last seen timestamp. Only push unrecorded events.
+	resp, err := client.LastSeen(ctx, &pb.LastSeenRequest{Name: r.Name})
```

Comment:
So do I understand correctly that "last seen" timestamp is stored on master nodes and indicates the last time they received an event from a particular node?

Reply:
The 'last seen' is the timestamp of the latest recorded event for a particular node. Will update doc to clarify.